Replace pointers to cipher_t/digest_t in connection_t with structs.

hg opened the pull request:
This is a part of #294 (replacing pointers to `cipher_t` and `digest_t` inside `connection_t` with structs themselves).

CI for this PR will fail until #321 is merged and this rebased on top.

I am not sure about changes to gcrypt. It is impossible to test because it hasn't been working for quite a while. I'll look into fixing it next.

gsliepen commented on src/openssl/prf.c:
```diff
@@ -30,14 +30,15 @@
 
 static bool prf_xor(int nid, const uint8_t *secret, size_t secretlen, uint8_t *seed, size_t seedlen, uint8_t *out, size_t outlen) {
-	digest_t *digest = digest_open_by_nid(nid, DIGEST_ALGO_SIZE);
+	digest_t *digest = digest_alloc();
```
This one can be allocated on the stack now.

gsliepen commented on src/protocol_auth.c:
```diff
@@ -509,18 +509,25 @@ bool send_metakey(connection_t *c) {
 
-	c->outbudget = cipher_budget(c->outcipher);
+	if(!cipher_open_by_name(&c->outcipher, cipher_name)) {
+		logger(DEBUG_ALWAYS, LOG_ERR, "Could not open cipher %s\n", cipher_name);
```
`cipher_open_by_name()` already logs an error, so we don't need this one.

hg commented:
Thank you, fixed.

https://github.com/hg/tinc/actions/runs/1133263859

I moved cross-compilation jobs to buster for now. They don't work on bullseye anymore. There have been some major changes to supported architectures IIRC. I'll look into it tomorrow.

It's a tiny change and does not warrant a separate pull request IMHO.

gsliepen commented:
Ooooh all green!
